File: simulations/domjudge/ContestSimulation.scala
package domjudge
import scala.concurrent.duration._

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import io.gatling.jdbc.Predef._
import scala.util.Random

import User._
import Team._
import Spectator._

class ContestSimulation extends Simulation {
  // Configure the base url, and other global options related to all requests
  val httpProtocol = http
    .baseUrl(System.getProperty("baseurl"))
     // This makes it load dependent resources: things like images, javascript, css, etc
    .inferHtmlResources()

  // Scenario that just grabs the public scoreboard every 30 seconds, some number of times
  // This will run for 10 minutes per "user". The scoreboard page autorefreshes every 30s,
  // so this is a good approximation of their activity.
  val spectatorScenario = scenario("PublicScoreboards").exec(Spectator.monitor_scoreboard(10))

  // Generate an infinite number of users, starting from gatling000000
  // var feeder = Iterator.from(0).map(i => Map("user" -> f"gatling$i%05d"))
  var feeder = tsv("accounts.tsv").eager.random

  // This scenario pretends to be a team. It takes around 3 minutes, and makes ~250 requests(including dependent resources)
  // It tries to be similar to what a team might do, load the team page, submit clarifications, submit problems, load the scoreboard, etc.
  val teamScenario = scenario("TeamExample")
    .feed(feeder)
    // Load the public scoreboard, then register/login
    .exec(Spectator.getscoreboard).pause(1)
    // Try a few times to get us registered/logged in, if it fails the user is done
    .tryMax(5) {
        exec(User.register(session => session("user").as[String], session => session("pass").as[String]))
        .exec(User.login(session => session("user").as[String], session => session("pass").as[String])).pause(2)
    }
    .exitHereIfFailed

    // Make a submission, view team information, check scoreboard
    .exec(Team.teampage).pause(5)
    .exec(Team.submit_java).pause(10)
    .exec(Team.teamscoreboard).pause(10)

    // Do some more team like stuff
    .exec(Team.teamdetails).pause(5)
    .exec(Team.teampage).pause(7)
    .exec(Team.teamscoreboard).pause(9)
    .exec(Team.teampage).pause(3)

    // Submit a clarification and then "check" it
    .exec(Team.requestclarification).pause(6)
    .exec(Team.teampage).pause(12)

    // Make some submissions, and check the team page periodically
    .exec(Team.teampage).pause(3)
    .exec(Team.teamscoreboard).pause(3)
    .exec(Team.submit_c).pause(15)
    .exec(Team.teampage).pause(6)
    .exec(Team.teampage).pause(14)
    .exec(Team.submit_py2).pause(8)
    .exec(Team.submit_hs).pause(5)
    .exec(Team.submit_kt).pause(5)
    .exec(Team.teampage).pause(3)
    .exec(Team.teampage).pause(4)
    .exec(Team.teamscoreboard).pause(8)
    .exec(Team.teampage).pause(2)
    .exec(Team.submit_nonewline).pause(11)
    .exec(Team.teampage).pause(16)
    .exec(Team.teampage).pause(5)

  setUp(teamScenario.inject(
      atOnceUsers(1) // start with 1 users
			// atOnceUsers(100) // start with 100 users
            //constantUsersPerSec(5) during (30 seconds) randomized  // gets to ~150 users in 30seconds
      //      constantUsersPerSec(18) during (60 seconds) randomized  // gets to ~1000(ish) users in 60seconds
			//rampUsers(500) over (60 seconds)

      // Use with the throttle commands below
      //constantUsersPerSec(30) during (45 seconds) // 1350 users?
  ), spectatorScenario.inject(atOnceUsers(1)))
    /*.throttle(
      jumpToRps(20).holdFor(2 minutes)
    )*/
    .maxDuration(6 minutes)
    .throttle(
      // reachRps(1) in (30 seconds),
      // reachRps(5) in (30 seconds),
      // reachRps(10) in (30 seconds),
      // reachRps(25) in (30 seconds),
      reachRps(50) in (30 seconds),
      holdFor(1 minute),
      reachRps(100) in (30 seconds),
      // holdFor(1 minute),
      // reachRps(150) in (30 seconds),
      // holdFor(1 minute),
      // reachRps(200) in (30 seconds),
      // holdFor(2 minutes)
      // reachRps(250) in (30 seconds),
      // holdFor(1 minute),
      // reachRps(300) in (30 seconds),
      // holdFor(1 minute),
      // reachRps(400) in (30 seconds),
      // holdFor(1 minute),
      // reachRps(500) in (30 seconds),
      // holdFor(5 minutes)
    )
    .protocols(httpProtocol)
    .pauses(uniformPausesPlusOrMinusPercentage(25)) // adjust pauses by 25percent either way

}

File: simulations/ContestSimulation.java
package domjudge;

import io.gatling.javaapi.core.*;
import io.gatling.javaapi.http.*;
import java.time.Duration;
import java.io.*;
import java.util.*;
import java.util.stream.Stream;
import java.util.function.*;
import java.util.concurrent.atomic.AtomicInteger;

import static io.gatling.javaapi.core.CoreDsl.*;
import static io.gatling.javaapi.http.HttpDsl.*;


public class ContestSimulation extends Simulation {
    // A fixed password (see seeding/accounts.json)
    String PASSWORD="secret";

    // How many "actions" a team will take (actions are things like submitting, requesting clarification, viewing a scoreboard)
    int NUM_TEAM_ACTIONS = 5;

    // Maximum amount of time the scenario can run
    int MAX_MINUTES = 5;

    private ChainBuilder login() {
        return exec(
            http("Login page get csrf")
            .get("/login")
            .check(
                css("input[name='_csrf_token']", "value")
                .find()
                .saveAs("csrftoken")
            ))
        .exec(http("Login Request")
            .post("/login")
            .formParam("_username", "#{user}")
            .formParam("_password", PASSWORD)
            .formParam("_csrf_token", "#{csrftoken}")
        );
    }

    private ChainBuilder submit(String langid, String filename) {
        return exec(
            http("Get submit form csrf from team page")
            .get("/team/submit")
            .check(
                css("input[id='submit_problem__token']", "value").find()
                .saveAs("csrftoken"),
                regex("<option value=\"([^\"]*)\">hello").find()
                .saveAs("problem_id")
            )
        ).exec(
            http(String.format("Submit solution %s", langid))
            .post("/team/submit")
            .formParam("submit_problem[_token]","#{csrftoken}")
            .formParam("submit_problem[problem]","#{problem_id}")
            .formParam("submit_problem[language]", langid)
            .formUpload("submit_problem[code][]", filename)
            .formParam("submit", "")
        );
    }

    private ChainBuilder requestClarification() {
        return exec(
            http("Get request clarification form")
            .get("/team/clarifications/add")
            .check(
                css("input[id=team_clarification__token", "value").find()
                .saveAs("csrftoken"),
                // get and save the clarification subject id for the "General issue" category
                regex("<option value=\"([^\"]*-general)\">General").find()
                .saveAs("clarification_subject")
            )
        ).exec(
            http("Request Clarification")
            .post("/team/clarifications/add")
            .formParam("team_clarification[recipient]", "dummy")
            .formParam("team_clarification[subject]", "#{clarification_subject}")
            .formParam("team_clarification[message]", "#{user} needs help")
            .formParam("team_clarification[_token]", "#{csrftoken}")
            .formParam("submit", "")
        );
    }

    /*********************** Spectator Scenario ****************************************************************
     * Scenario that just grabs the public scoreboard every 30 seconds. The scoreboard page autorefreshes
     * every 30s, so this is a good approximation of their activity.
     */
    HttpRequestActionBuilder publicScoreboard = http("Public Scoreboard").get("/public");
    ChainBuilder monitorScoreboard = exec(poll().every(30).exec(publicScoreboard)).pause(Duration.ofMinutes(MAX_MINUTES));
    ScenarioBuilder spectatorScenario = scenario("SpectatorScenario").exec(monitorScoreboard);


    /*********************** Team Scenario *********************************************************************
     * This pretends to be a team, it polls the team page every 30s and randomly performs various actions
     */
    // Feeder for team accounts
    AtomicInteger counter = new AtomicInteger(1);
    Iterator<Map<String,Object>> feeder =
        Stream.generate((Supplier<Map<String,Object>>) () -> {
            String username = String.format("gatling%04d", counter.getAndIncrement());
            return Collections.singletonMap("user", username);
        }
    ).iterator();

    HttpRequestActionBuilder teamPage = http("Team Page").get("/team");
    ChainBuilder teamSubmitC = submit("c", "test-hello.c");
    ChainBuilder teamSubmitCpp = submit("cpp", "test-hello.c++");
    ChainBuilder teamSubmitJava = submit("java", "test-hello.java");
    ChainBuilder teamSubmitKotlin = submit("kt", "test-hello.kt");
    ChainBuilder teamSubmitPython = submit("py3", "test-hello.py3");
    ChainBuilder teamRequestClarification = requestClarification();
    ChainBuilder teamProblems = exec(http("Team Problemset").get("/team/problems"));
    ChainBuilder teamScoreboard = exec(http("Team Scoreboard").get("/team/scoreboard"));
    ScenarioBuilder teamScenario = scenario("TeamScenario")
        .feed(feeder)
        .exec(http("Homepage").get("/"))
        .exec(publicScoreboard) // Fetch the public scoreboard
        .pause(1,5)               // Brief delay
        .exec(login())            // Log in as the team

        // Mimic the behavior of the team page, it autorefreshes every 30s
        .exec(
            poll().pollerName("teampage")
            .every(30)
            .exec(teamPage)
        )

        // While we're polling the teampage, lets pretend to do some actions like a real team might
        .repeat(NUM_TEAM_ACTIONS, "i").on(
            exec(
                randomSwitch().on(
                    // 20% chance of making a submission
                    Choice.withWeight(20.0, exec(
                        uniformRandomSwitch().on(
                            exec(teamSubmitC),
                            exec(teamSubmitCpp),
                            exec(teamSubmitJava),
                            exec(teamSubmitKotlin),
                            exec(teamSubmitPython)
                        )
                    )),
                    // 10% of requesting clarification
                    Choice.withWeight(10.0, teamRequestClarification),
                    // 30% of loading problems page
                    Choice.withWeight(30.0, teamProblems),
                    // 40% of loading scoreboard
                    Choice.withWeight(40.0, teamScoreboard)
                )
            )
            // Sleep anywhere between 5s and 1 minute before doing another action
            .pause(5,60)
        )

        // Stop polling the teampage
        .exec(
            poll().pollerName("teampage").stop()
        )

    ;


    HttpProtocolBuilder httpProtocol = http
        .baseUrl(System.getProperty("baseurl"))
        .inferHtmlResources()   // load all dependent things like images/js/css/etc
        .nameInferredHtmlResourcesAfterPath()
    ;

    {
        setUp(
            // teamScenario.injectOpen(atOnceUsers(5)),
            teamScenario.injectClosed(
                incrementConcurrentUsers(10)  // Batches of this many users
                .times(10)                    // how many "levels" to test
                .eachLevelLasting(60)         // hold the number of users for this duration
                .separatedByRampsLasting(60)  // Take 60s between adding each user
                .startingFrom(10)             // Start with 10 users
            ),
            // spectatorScenario.injectOpen(atOnceUsers(10))
            spectatorScenario.injectClosed(
                incrementConcurrentUsers(10)  // Batches of this many spectators
                .times(10)                    // how many "levels" to test
                .eachLevelLasting(60)         // hold the number of users for this duration
                .separatedByRampsLasting(60)  // Take 60s between adding each user
                .startingFrom(10)             // Start with 10 users
            )
        )
        .maxDuration(Duration.ofMinutes(MAX_MINUTES))
        .protocols(httpProtocol);
    }
}
